src/storage/pack.rs

```diff
@@ -0,0 +1,388 @@
+        tar_append_file(store, tar, layer, &path, f, mtime).await?;
+    }
+    for f in &SHARED_OPTIONAL_FILES {
+        if f == &FILENAMES.rollup {
+            // skip the rollup file. It will not be resolvable remotely.
+            continue;
+        }
+        tar_append_file_if_exists(store, tar, layer, &path, f, mtime).await?;
+    }
+    if store.file_exists(layer, FILENAMES.parent).await? {
+        // this is a child layer
+        for f in &CHILD_LAYER_REQUIRED_FILES {
+            tar_append_file(store, tar, layer, &path, f, mtime).await?;
+        }
+        for f in &CHILD_LAYER_OPTIONAL_FILES {
+            tar_append_file_if_exists(store, tar, layer, &path, f, mtime).await?;
+        }
+    } else {
+        // this is a base layer
+        for f in &BASE_LAYER_REQUIRED_FILES {
+            tar_append_file(store, tar, layer, &path, f, mtime).await?;
+        }
+        for f in &BASE_LAYER_OPTIONAL_FILES {
+            tar_append_file_if_exists(store, tar, layer, &path, f, mtime).await?;
+        }
+    }
+
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum PackError {
+    LayerNotFound,
+    Io(io::Error),
+    Utf8Error(std::str::Utf8Error),
+}
+
+impl Display for PackError {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        write!(formatter, "{:?}", self)
+    }
+}
+
+impl From<io::Error> for PackError {
+    fn from(err: io::Error) -> Self {
+        Self::Io(err)
+    }
+}
+impl From<std::str::Utf8Error> for PackError {
+    fn from(err: std::str::Utf8Error) -> Self {
+        Self::Utf8Error(err)
+    }
+}
+
+pub fn pack_layer_parents<R: io::Read>(
+    readable: R,
+) -> Result<HashMap<[u32; 5], Option<[u32; 5]>>, PackError> {
+    let tar = GzDecoder::new(readable);
+    let mut archive = Archive::new(tar);
+
+    // build a set out of the layer ids for easy retrieval
+    let mut result_map = HashMap::new();
+
+    println!("about to go into loop");
+    for e in archive.entries()? {
+        println!("a");
+        let mut entry = e?;
+        println!("b");
+        let path = entry.path()?;
+        println!("c");
+        println!("path: {:?}", path);
+
+        let id = string_to_name(
+            path.iter()
+                .next()
+                .expect("expected path to have at least one component")
+                .to_str()
+                .expect("expected proper unicode path"),
+        )?;
+
+        if path.file_name().expect("expected path to have a filename") == "parent.hex" {
+            println!("hello a");
+            // this is an element we want to know the parent of
+            // lets read it
+            let mut parent_id_bytes = [0u8; 40];
+            entry.read_exact(&mut parent_id_bytes)?;
+            let parent_id_str = std::str::from_utf8(&parent_id_bytes)?;
+            let parent_id = string_to_name(parent_id_str)?;
+
+            result_map.insert(id, Some(parent_id));
+        } else if !result_map.contains_key(&id) {
```

Comment:
https://rust-lang.github.io/rust-clippy/master/index.html#map_entry

contains_key followed by insert is inefficient